pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerManager.java

```diff
@@ -502,12 +506,18 @@ private void assignWorkersToNonPartitionedLeafFragment(DispatchablePlanMetadata
         metadata.addUnavailableSegments(tableName, routingTable.getUnavailableSegments());
       }
     }
+    // Sort server instances to ensure deterministic worker ID assignment.
+    // This is critical for pre-partitioned exchanges where worker ID N on one stage
+    // must map to the same physical server as worker ID N on another stage.
+    List<ServerInstance> sortedServers = new ArrayList<>(serverInstanceToSegmentsMap.keySet());
+    sortedServers.sort(Comparator.comparing(ServerInstance::getInstanceId));
+
     int workerId = 0;
     Map<Integer, QueryServerInstance> workerIdToServerInstanceMap = new HashMap<>();
     Map<Integer, Map<String, List<String>>> workerIdToSegmentsMap = new HashMap<>();
-    for (Map.Entry<ServerInstance, Map<String, List<String>>> entry : serverInstanceToSegmentsMap.entrySet()) {
-      workerIdToServerInstanceMap.put(workerId, new QueryServerInstance(entry.getKey()));
-      workerIdToSegmentsMap.put(workerId, entry.getValue());
+    for (ServerInstance serverInstance : sortedServers) {
```

Comment:
(minor) Do a for loop on `numWorkers = sortedServers.size()` is more intuitive after the change